### DISCUSSION TITLE
Triggering downstream pipeline when all dependency pipelines complete

### DISCUSSION THREAD

**Discussion by IAXES** [concourse/concourse#8206] [Help & Support] [answered]
# Background

I have a number of applications (3 for now, but I expect the number to grow with time). Each application is maintained by a different team, and each application has its own Concourse build pipeline maintained and owned by the associated team. The key piece of each application's pipeline of note is that it:

1. Pulls in a common dependency (communications library).
2. Builds the "business logic" application against this library.
3. Packages the final application plus the library (and other artifacts) into a Docker container, and publishes this image to a Docker registry.
4. All the resulting container images are bundled into a VM image (maintained collectively by all three teams). An ASCII art render of the overall project structure is shown below.

```
         VM Image       
           /|\          
          / | \         
         /  |  \        
        /   |   \       
       /    |    \      
      /     |     \     
     /      |      \    
    /       |       \   
 C.Img.   C.Img.  C.Img.
 Alpha    Beta    Gamma 
   |        |        |  
 Build    Build   Build 
 Alpha    Beta    Gamma 
   |        |        |  
    \       |       /   
     \      |      /    
      \     |     /     
       \    |    /      
        \   |   /       
         \  |  /        
          \ | /         
           \|/          
         Common         
          Deps          
```

# Problem

I need to ensure that all three dependencies (or `n` dependencies in the future, as this scales to include more applications) are using the same API version of the "common deps" artifacts (which are produced by an independent build pipeline), to guarantee that the final `Alpha`, `Beta`, and `Gamma` containerized applications are guaranteed to be version-aligned (i.e. API compatible) with each other, and prevent avoidable issues due to API-mismatched artifacts finding their way into the final VM image.

Additionally, I need to be able to configure the overall pipeline so that, if a new build of `Common Deps` is triggered, it triggers builds of the `Alpha`, `Beta`, and `Gamma` pipelines, and then triggers a build of `VM Image`, but **only** when all three versioned-aligned builds of `Alpha`, `Beta`, and `Gamma` are ready (i.e. guarantee we don't encounter build-time race conditions resulting in API-mismatched copies of the three dependencies being bundled into the final `VM Image`). So, once some important trigger/pull-request/button-press is executed, the entire "build common deps + version-aligned `Alpha`/`Beta`/`Gamma` + bundle them into `VM Image`" is completely automated.

Lastly, if `Common Deps` is not updated, but the one of the individual `Alpha`/`Beta`/`Gamma` pipelines is triggered, `VM Image` should be triggered  as well. 

# Constraints

Teams need to retain control of their own pipelines: we can't refactor these three projects into a single pipeline/YAML file, as it would be an anti-pattern as far as scaling teams goes. We can add things like external resources such as `semver` objects with `trigger:true` configured for them in the various pipelines, but a major refactor is a no-go.

# Thoughts So Far

I can use various resources such as a `semver` to trigger the re-builds of `Alpha`, `Beta`, and `Gamma` easily enough. The part requiring more thought is how to configure `VM Image` to only build when version-aligned (i.e. via some simplistic dependency matrix) releases of all three dependencies are available. I could have the build job itself (i.e. some pre-flight-check/sanity-check script) execute this check, perhaps via a polling loop approach, but I'd like to check with the community if there's a more canonical or "clean" approach to this.

I [considered parameterized jobs, but based on what's available](https://github.com/concourse/concourse/issues/783), I don't think this could be accomplished without some hacky'ish build scripts running in the jobs (or having a job call `fly trigger-job --watch` itself, which seems to be an anti-pattern).

I also reviewed [a ticket on combined/aggregated triggers](https://github.com/concourse/concourse/issues/2324#issuecomment-461332192), but am still trying to determine how to make this work with separate independent pipelines (i.e. can't refactor our pipelines into jobs).

Lastly, I'm thinking of some sort of "cache/blob" build job, which includes some simple scripts to determine whether or not to trigger the `VM Image` job (ASCII art render below). It would work, but still seems like a bit of a hack, as it would be implementing a mini-CI-system of sorts, into a broader CI system. On the plus side, it wouldn't be hard for that system to do a check along the lines of "do `Alpha`, `Beta`, and `Gamma` all use the same version of `Common Deps`" before allowing the build to proceed further. Sort like a passive parameter check that lets us approximate some of the functionality that parameterized jobs would provide. Even with this, there are some build-time race conditions that come to mind.

```
         VM Image       
            |
       Cache/Blob +
     Dep. Alignment
            |
           /|\          
          / | \         
         /  |  \        
        /   |   \       
       /    |    \      
      /     |     \     
     /      |      \    
    /       |       \   
 C.Img.   C.Img.  C.Img.
 Alpha    Beta    Gamma 
   |        |        |  
   ```

# Question

Any comments/suggestions/recommendations on how to best proceed with such a pipeline automation setup w/ Concourse?

Thank you.

**Comment by suhlig** [comment]
I think you could configure the pipeline that builds `VM Image` to `get` an _exact_ version of the common dependency (e.g. `v1.0.1`). Whenever the common dependency was bumped (e.g. to `v1.0.2`), a `set_pipeline` step could update the version that is defined in the pipeline that builds `VM Image` (e.g. by using a `load_var` step). 

I do this all the time with the help of the [`github-release-resource`](https://github.com/concourse/github-release-resource): When a new version of pandoc was released, a task image that installs pandoc as a dependency is re-built and pushed again. The new pandoc version is read from a file by `load_var` and passed as build argument to the image build.

**Reply by IAXES** [reply]
I experimented with a similar approach. What I originally had was:

- `common-deps` updates a semver in an S3 bucket.
- This triggers the `alpha`, `beta`, and `gamma` builds.
- These each trigger their respective container builds when done.

At this point, we hit a CI pipeline race condition. Any one of the `alpha`/`beta`/`gamma` container builds can trigger the final VM image build, but there's no way to combine (i.e. similar to a thread barrier or timed gate) the three builds, so that the VM image build triggers when all three are simultaneously available (unless we write an intermediate pipeline to handle this corner case, which seemed absurd).

So, from here, I eventually discovered I could:

- Have the VM image run on a timer, and check if new dependencies were available. However, due to a lack of `if-else` functionality in Concourse, this meant that we have to replace if-else logic with pass/fail logic (and build failures result in notifications; not good). Also, having to trigger an async job on a timer defeated the purpose of an async job (i.e. avoiding resources being wasted).
- Manually triggering builds when we knew all dependencies were available (but then we lose automaton; a major show stopper).
- Have an external CI system (i.e. tried out Jenkins and GitLab) handle this "glue logic" ("why do we even have Concourse if we're offloading complex tasks to another product" arose from that experiment).
- Combine all our build pipelines from different teams into a single pipeline (also a show-stopper; anti-pattern as far as scaling goes).

Ultimately, if we wanted to avoid wasting resources, have the jobs 100% automated, etc.; we had to move to Jenkins or GitLab or some other tool that supports:

- Parameterized builds (i.e. real parameterized and automated builds; manually overriding pipeline variables using `fly` CLI doesn't count; please see https://github.com/concourse/concourse/issues/783).
- Conditional statements.

A friend was able to throw this together in about 10 minutes on **both** Jenkins and GitLab, so we just arrived at the conclusion that Concourse is just not the tool for us.

**Comment by IAXES** [comment, marked as answer]
Closing. Thanks for the help.

TLDR: one does not implement this in Concourse.
